msvc 2015 warnings

The code below doesn't compile with the newest winapi modules and /WX

`
boost::process::child c("c:/windows/system32/foo.exe", "/r", dll);
c.wait();
`

Visual Studio 2015
64-bit debug build
boost 1.62.0 static build, with static runtime
flags: /WX /W4 
pp: _SCL_SECURE_NO_WARNINGS;_CRT_SECURE_NO_WARNINGS;

`
Severity	Code	Description	Project	File	Line	Suppression State
Warning	C4244	'argument': conversion from '__int64' to 'int', possible loss of data	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\locale.hpp	49	
Warning	C4244	'argument': conversion from '__int64' to 'int', possible loss of data	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\locale.hpp	71	
Warning	C4800	'boost::detail::winapi::BOOL_': forcing value to bool 'true' or 'false' (performance warning)	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\child_handle.hpp	85	
Warning	C4800	'boost::detail::winapi::BOOL_': forcing value to bool 'true' or 'false' (performance warning)	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\child_handle.hpp	92	
Warning	C4800	'boost::detail::winapi::BOOL_': forcing value to bool 'true' or 'false' (performance warning)	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\group_handle.hpp	146	
Warning	C4800	'boost::detail::winapi::BOOL_': forcing value to bool 'true' or 'false' (performance warning)	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\group_handle.hpp	153	
Warning	C4800	'boost::detail::winapi::BOOL_': forcing value to bool 'true' or 'false' (performance warning)	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\group_handle.hpp	183	
Warning	C4099	'boost::process::detail::windows::group_handle': type name first seen using 'struct' now seen using 'class'	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\wait_group.hpp	16	
Warning	C4100	'state': unreferenced formal parameter	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\locale.hpp	38	
Warning	C4100	'state': unreferenced formal parameter	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\locale.hpp	60	
d:\src\trunk\third_party\boost\include\boost\fusion\container\generation\detail\preprocessed\make_set10.hpp	35	
Warning	C4459	declaration of 'env' hides global declaration	fpreg	D:\src\trunk\third_party\boost\include\boost\process\environment.hpp	57	
Warning	C4458	declaration of 'ios' hides class member	fpreg	D:\src\trunk\third_party\boost\include\boost\process\async.hpp	59	
Warning	C4459	declaration of 'on_error' hides global declaration	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\executor.hpp	183	
Warning	C4459	declaration of 'on_error' hides global declaration	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\executor.hpp	215	
Warning	C4459	declaration of 'on_setup' hides global declaration	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\executor.hpp	178	
Warning	C4459	declaration of 'on_success' hides global declaration	fpreg	D:\src\trunk\third_party\boost\include\boost\process\detail\windows\executor.hpp	206

Thanks for the actual compiler output, precise error reports do in fact help.

Now to clarify my point of view: it makes a lot of sense for a header-only library to minimize the amount of warnings generated. And there's room for improvement here, though it's not high on my priority list. You can of course accelerate that by providing a PR which I'll gladly accept.
It will however not be possible to provide you a warning free library, since I will neither put a plain "#pragma" into the code (which will generate a warning on another compiler) nor will I allow any performance hit.

I do think, that warning-as-error is a bad policy, especially with stupid warnings like C4458 or worse C4459.
There are in fact other boost libraries generating warnings. I've built it often enough to know, for example, gcc will always generate this warning when including `boost.asio`:

```cpp
../../../boost/asio/error.hpp:258:45: warning: ‘boost::asio::error::system_category’ defined but not used [-Wunused-variable]
 static const boost::system::error_category& system_category
```

There are just things the compiler complains about, but it's not bad, not an error, not a sign of bad code. Having no warnings is no sign of good code, having them not a sign of bad code. I feel for you, when you have this policy at work, but please understand why it's not as important for me as it is for you.

--- a/include/boost/process/async.hpp
+++ b/include/boost/process/async.hpp
@@ -71,7 +71,7 @@ struct async_builder
 {
     boost::asio::io_service * ios;
 
-    void operator()(boost::asio::io_service & ios) {this->ios = &ios;};
+    void operator()(boost::asio::io_service & ios_) {this->ios = &ios_;};
 
     typedef api::io_service_ref result_type;
     api::io_service_ref get_initializer() {return api::io_service_ref (*ios);};

--- a/include/boost/process/detail/windows/child_handle.hpp
+++ b/include/boost/process/detail/windows/child_handle.hpp
@@ -82,14 +82,14 @@ struct child_handle
         ::boost::detail::winapi::BOOL_ value;
         if (!::boost::detail::winapi::IsProcessInJob(proc_info.hProcess, nullptr, &value))
             throw_last_error("IsProcessinJob Failed");
-        return value;
+        return value==1; // TODO: correct Windows API usage is: value==TRUE but TRUE isn't defined at winapi
     }
     bool in_group(std::error_code &ec) const noexcept
     {
         ::boost::detail::winapi::BOOL_ value;
         if (!::boost::detail::winapi::IsProcessInJob(proc_info.hProcess, nullptr, &value))
             ec = get_last_error();
-        return value;
+        return value==1;
     }
 };
 

--- a/include/boost/process/detail/windows/child_handle.hpp
+++ b/include/boost/process/detail/windows/child_handle.hpp
@@ -82,14 +82,14 @@ struct child_handle
         ::boost::detail::winapi::BOOL_ value;
         if (!::boost::detail::winapi::IsProcessInJob(proc_info.hProcess, nullptr, &value))
             throw_last_error("IsProcessinJob Failed");
-        return value==1; // TODO: correct Windows API usage is: value==TRUE but TRUE isn't defined at winapi
+        return value!=0;
     }
     bool in_group(std::error_code &ec) const noexcept
     {
         ::boost::detail::winapi::BOOL_ value;
         if (!::boost::detail::winapi::IsProcessInJob(proc_info.hProcess, nullptr, &value))
             ec = get_last_error();
-        return value==1;
+        return value!=0;
     }
 };
 

--- a/include/boost/process/detail/windows/executor.hpp
+++ b/include/boost/process/detail/windows/executor.hpp
@@ -175,13 +175,13 @@ public:
 
     child operator()()
     {
-        on_setup_t on_setup(*this);
-        boost::fusion::for_each(seq, on_setup);
+        on_setup_t on_setup_fn(*this);
+        boost::fusion::for_each(seq, on_setup_fn);
 
         if (_ec)
         {
-            on_error_t on_error(*this, _ec);
-            boost::fusion::for_each(seq, on_error);
+            on_error_t on_error_fn(*this, _ec);
+            boost::fusion::for_each(seq, on_error_fn);
             return child();
         }
 
@@ -203,8 +203,8 @@ public:
         if (err_code != 0)
         {
             _ec.clear();
-            on_success_t on_success(*this);
-            boost::fusion::for_each(seq, on_success);
+            on_success_t on_success_fn(*this);
+            boost::fusion::for_each(seq, on_success_fn);
         }
         else
             set_error(::boost::process::detail::get_last_error(),
@@ -212,8 +212,8 @@ public:
 
         if ( _ec)
         {
-            on_error_t on_error(*this, _ec);
-            boost::fusion::for_each(seq, on_error);
+            on_error_t on_err(*this, _ec);
+            boost::fusion::for_each(seq, on_err);
             return child();
         }
         else

--- a/include/boost/process/detail/windows/group_handle.hpp
+++ b/include/boost/process/detail/windows/group_handle.hpp
@@ -143,14 +143,14 @@ struct group_handle
         if (!::boost::detail::winapi::IsProcessInJob(proc, _job_object,  &is))
             throw_last_error();
 
-        return is;
+        return is==1;
     }
     bool has(handle_t proc, std::error_code & ec) noexcept
     {
         ::boost::detail::winapi::BOOL_ is;
         if (!::boost::detail::winapi::IsProcessInJob(proc, _job_object,  &is))
             ec = get_last_error();
-        return is;
+        return is==1;
     }
 
     bool valid() const
@@ -180,7 +180,7 @@ inline bool in_group()
     if (!::boost::detail::winapi::IsProcessInJob(boost::detail::winapi::GetCurrentProcess(), nullptr, &res))
         throw_last_error("IsProcessInJob failed");
 
-    return res;
+    return res==1;
 }
 
 

--- a/include/boost/process/detail/windows/group_handle.hpp
+++ b/include/boost/process/detail/windows/group_handle.hpp
@@ -143,14 +143,14 @@ struct group_handle
         if (!::boost::detail::winapi::IsProcessInJob(proc, _job_object,  &is))
             throw_last_error();
 
-        return is==1;
+        return is!=0;
     }
     bool has(handle_t proc, std::error_code & ec) noexcept
     {
         ::boost::detail::winapi::BOOL_ is;
         if (!::boost::detail::winapi::IsProcessInJob(proc, _job_object,  &is))
             ec = get_last_error();
-        return is==1;
+        return is!=0;
     }
 
     bool valid() const
@@ -180,7 +180,7 @@ inline bool in_group()
     if (!::boost::detail::winapi::IsProcessInJob(boost::detail::winapi::GetCurrentProcess(), nullptr, &res))
         throw_last_error("IsProcessInJob failed");
 
-    return res==1;
+    return res!=0;
 }
 
 

--- a/include/boost/process/detail/windows/locale.hpp
+++ b/include/boost/process/detail/windows/locale.hpp
@@ -8,6 +8,7 @@
 #define BOOST_PROCESS_DETAIL_WINDOWS_LOCALE_HPP_
 
 #include <locale>
+#include <boost/core/ignore_unused.hpp>
 #include <boost/detail/winapi/file_management.hpp>
 #include <boost/detail/winapi/character_code_conversion.hpp>
 
@@ -39,6 +40,7 @@ class windows_file_codecvt
      const char* from, const char* from_end, const char*& from_next,
      wchar_t* to, wchar_t* to_end, wchar_t*& to_next) const override
    {
+     boost::ignore_unused(state);
      ::boost::detail::winapi::UINT_ codepage = AreFileApisANSI() ?
              ::boost::detail::winapi::CP_ACP_ :
              ::boost::detail::winapi::CP_OEMCP_;
@@ -46,7 +48,7 @@ class windows_file_codecvt
      int count;
      if ((count = ::boost::detail::winapi::MultiByteToWideChar(codepage,
              ::boost::detail::winapi::MB_PRECOMPOSED_, from,
-       from_end - from, to, to_end - to)) == 0)
+       static_cast<int>(from_end - from), to, static_cast<int>(to_end - to))) == 0)
      {
        return error;  // conversion failed
      }
@@ -61,6 +63,7 @@ class windows_file_codecvt
      const wchar_t* from, const wchar_t* from_end, const wchar_t*& from_next,
      char* to, char* to_end, char*& to_next) const override
    {
+     boost::ignore_unused(state);
      auto codepage = ::boost::detail::winapi::AreFileApisANSI() ?
                        ::boost::detail::winapi::CP_ACP_ :
                      ::boost::detail::winapi::CP_OEMCP_;
@@ -68,7 +71,7 @@ class windows_file_codecvt
      int count;
      if ((count = ::boost::detail::winapi::WideCharToMultiByte(codepage,
                    ::boost::detail::winapi::WC_NO_BEST_FIT_CHARS_, from,
-                  from_end - from, to, to_end - to, 0, 0)) == 0)
+                  static_cast<int>(from_end - from), to, static_cast<int>(to_end - to), 0, 0)) == 0)
      {
        return error;  // conversion failed
      }

--- a/include/boost/process/detail/windows/wait_group.hpp
+++ b/include/boost/process/detail/windows/wait_group.hpp
@@ -13,7 +13,7 @@
 
 namespace boost { namespace process { namespace detail { namespace windows {
 
-class group_handle;
+struct group_handle;
 
 inline void wait(const group_handle &p)
 {

--- a/include/boost/process/environment.hpp
+++ b/include/boost/process/environment.hpp
@@ -54,8 +54,8 @@ struct const_entry
             return string_type();
     }
     string_type get_name() const {return string_type(_name.begin(), _name.end());}
-    explicit const_entry(string_type&& name, pointer data, environment_t & env) :
-        _name(std::move(name)), _data(data), _env(&env) {}
+    explicit const_entry(string_type&& name, pointer data, environment_t & env_) :
+        _name(std::move(name)), _data(data), _env(&env_) {}
 
     explicit const_entry(string_type &&name, environment_t & env) :
         _name(std::move(name)), _data(nullptr), _env(&env) {}

--- a/test/Jamfile.jam
+++ b/test/Jamfile.jam
@@ -19,8 +19,6 @@ project : requirements
   <toolset>msvc:<define>_SCL_SECURE_NO_WARNINGS
   <toolset>msvc:<define>_CRT_SECURE_NO_DEPRECATE
   <target-os>windows:<define>WIN32_LEAN_AND_MEAN
-  <target-os>windows:<define>NOMINMAX
-  <target-os>windows:<define>_WIN32_WINNT=0x0501
   <target-os>linux:<linkflags>-lpthread
   <os>NT,<toolset>cw:<library>ws2_32
   <os>NT,<toolset>gcc:<library>ws2_32

--- a/test/Jamfile.jam
+++ b/test/Jamfile.jam
@@ -16,8 +16,11 @@ if [ os.name ] = NT
 }
 
 project : requirements
-  <toolset>msvc:<define>_SCL_5SECURE_NO_WARNINGS
+  <toolset>msvc:<define>_SCL_SECURE_NO_WARNINGS
+  <toolset>msvc:<define>_CRT_SECURE_NO_DEPRECATE
   <target-os>windows:<define>WIN32_LEAN_AND_MEAN
+  <target-os>windows:<define>NOMINMAX
+  <target-os>windows:<define>_WIN32_WINNT=0x0501
   <target-os>linux:<linkflags>-lpthread
   <os>NT,<toolset>cw:<library>ws2_32
   <os>NT,<toolset>gcc:<library>ws2_32
@@ -38,14 +41,14 @@ alias coroutine        : /boost//coroutine       : <link>static ;
 exe sparring_partner : sparring_partner.cpp program_options system filesystem iostreams :
     <warnings>off
     ;
-    
-exe sub_launch : sub_launcher.cpp program_options iostreams system filesystem : <warnings>off ;    
+
+exe sub_launch : sub_launcher.cpp program_options iostreams system filesystem : <warnings>off ;
 
 test-suite bare :
     [ run environment.cpp system filesystem  ]
     [ run pipe.cpp        system filesystem  ]
     ;
-    
+
 test-suite execution :
     [ run async.cpp       system thread filesystem            : : sparring_partner ]
     [ run async_fut.cpp   system thread filesystem            : : sparring_partner ]

